fast64_internal/f3d/f3d_writer.py

```diff
@@ -2024,6 +2027,53 @@ def normToSigned8Vector(normal):
 	return [int.from_bytes(int(value * 127).to_bytes(1, 'big',
 		signed = True), 'big') for value in normal]
 
+def saveBitGeoGBI2(value, defaultValue, flagName, geo, matWriteMethod):
+	if value != defaultValue or matWriteMethod == GfxMatWriteMethod.WriteAll:
+		if value:
+			geo.setFlagList.append(flagName)
+		else:
+			geo.clearFlagList.append(flagName)
+
+def saveGeoModeDefinitionGBI2(fMaterial, settings, defaults, matWriteMethod):
+	geo = SPGeometryMode([],[])
+
+	saveBitGeoGBI2(settings.g_zbuffer, defaults.g_zbuffer, 'G_ZBUFFER',
+		geo, matWriteMethod)
+	saveBitGeoGBI2(settings.g_shade, defaults.g_shade, 'G_SHADE',
+		geo, matWriteMethod)
+	saveBitGeoGBI2(settings.g_cull_front, defaults.g_cull_front, 'G_CULL_FRONT',
+		geo, matWriteMethod)
+	saveBitGeoGBI2(settings.g_cull_back,  defaults.g_cull_back, 'G_CULL_BACK',
+		geo, matWriteMethod)
+	saveBitGeoGBI2(settings.g_fog, defaults.g_fog, 'G_FOG', geo, matWriteMethod)
+	saveBitGeoGBI2(settings.g_lighting, defaults.g_lighting, 'G_LIGHTING',
+		geo, matWriteMethod)
+
+	# make sure normals are saved correctly.
+	saveBitGeoGBI2(settings.g_tex_gen, defaults.g_tex_gen, 'G_TEXTURE_GEN',
+		geo, matWriteMethod)
+	saveBitGeoGBI2(settings.g_tex_gen_linear, defaults.g_tex_gen_linear,
+		'G_TEXTURE_GEN_LINEAR', geo, matWriteMethod)
+	saveBitGeoGBI2(settings.g_shade_smooth, defaults.g_shade_smooth,
+		'G_SHADING_SMOOTH', geo, matWriteMethod)
+	if bpy.context.scene.f3d_type == 'F3DEX_GBI_2' or \
+		bpy.context.scene.f3d_type == 'F3DEX_GBI':
+		saveBitGeoGBI2(settings.g_clipping, defaults.g_clipping, 'G_CLIPPING',
+			geo, matWriteMethod)
+
+	if len(geo.clearFlagList) == 0:
+		geo.clearFlagList.append('0')
+	elif len(geo.setFlagList) == 0:
+		geo.setFlagList.append('0')
+
+	if matWriteMethod == GfxMatWriteMethod.WriteAll:
+		fMaterial.material.commands.append(SPLoadGeometryMode(geo.setFlagList))
+	else:
+		fMaterial.material.commands.append(geo)
+
+	if matWriteMethod == GfxMatWriteMethod.WriteDifferingAndRevert:
+		fMaterial.revert.commands.append(SPGeometryMode(geo.setFlagList, geo.clearFlagList))
```

Comment:
I'm pretty sure the only check on the mat write method that is needed is in `saveBitGeoGBI2`
```suggestion
	fMaterial.material.commands.append(geo)
```

Reply:
I don't think this would be equivalent. I'll explain:

I check if WriteAll is used first, since that means that all flags will either be set or cleared, so LoadGeometryMode can be used in all cases for that scenario, however if Differing and Revert is used then that is not the case. I am now realizing that I can still clean this code up a little bit here though.